lib/blueprinter/extension.rb

```diff
@@ -2,11 +2,195 @@
+  #      - prepare (only first time during a given render)
+  #      - blueprint_fields (only first time during a given render)
+  #      - blueprint_input
+  #      - field_value
+  #      - exclude_field?
+  #      - object_value
+  #      - exclude_object?
+  #      - collection_value
+  #      - exclude_collection?
+  #      - blueprint_output
+  #    - output_object | output_collection
+  #    - json
+  #
+  # V1 hook call order:
+  #  - pre_render
   #
   class Extension
     #
-    # Called eary during "render", this method receives the object to be rendered and
+    # Returns true if the given object should be treated as a collection (i.e. supports `map { |obj| ... }`).
+    #
+    # @param _object [Object]
+    # @return [Boolean]
+    #
+    def collection?(_object) = false
+
+    #
+    # Runs around the entire rendering process. MUST yield!
+    #
+    # @param _context [Blueprinter::V2::Context]
+    #
+    def around(_context) = yield
+
+    #
+    # Runs around serialization of a Blueprint object. Surrounds the `prepare` through `blueprint_output` hooks. MUST yield!
+    #
+    # @param _context [Blueprinter::V2::Context]
+    #
+    def around_object(_context) = yield
+
+    #
+    # Runs around serialization of a Blueprint collection. Surrounds the `prepare` through `blueprint_output` hooks. MUST yield!
+    #
+    # @param _context [Blueprinter::V2::Context]
+    #
+    def around_collection(_context) = yield
+
+    #
+    # Called once per blueprint per render. A common use is to pre-calculate certain options
+    # and cache them in context.data, so we don't have to recalculate them for every field.
+    #
+    # @param _context [Blueprinter::V2::Context]
+    #
+    def prepare(_context) = nil
+
+    #
+    # Returns the fields that should be included in the correct order. Default is all fields in the order in which they were defined.
+    #
+    # NOTE Only runs once per Blueprint per render.
+    #
+    # @param _context [Blueprinter::V2::Context]
+    # @return [Array<Blueprinter::V2::Field|Blueprinter::V2::Object|Blueprinter::V2::Collection>]
+    #
+    def blueprint_fields(_context) = []
+
+    #
+    # Modify or replace the object passed to render/render_object. The returned object is what will be rendered.
+    #
+    # @param context [Blueprinter::V2::Context]
+    # @return [Object]
+    #
+    def input_object(context) = context.object
+
+    #
+    # Modify or replace the collection passed to render/render_collection. The returned collection is what will be rendered.
+    #
+    # @param context [Blueprinter::V2::Context]
+    # @return [Object]
+    #
+    def input_collection(context) = context.object
+
+    #
+    # Modify or replace the object result (stored in context.value) before final render (e.g. to JSON). The returned object is what will be rendered.
+    #
+    # @param context [Blueprinter::V2::Context]
+    # @return [Object]
+    #
+    def output_object(context) = context.value
```

Comment:
This might be suffering from our use of serialize/render, but could we rephrase this as `after_serialization` / `post_serialization` or something similar?